encoder/v1/encoder.go

```diff
@@ -0,0 +1,218 @@
+		{Name: "linkNative", Type: chain.Uint256},
+		{Name: "upkeepIds", Type: chain.Uint256Arr},
+		{Name: "wrappedPerformDatas", Type: chain.PerformDataArr},
+	}
+
+	reportKeys := make([]string, len(arguments))
+	for i, arg := range arguments {
+		reportKeys[i] = arg.Name
+	}
+
+	return &encoder{
+		reportKeys: reportKeys,
+		packer:     arguments,
+	}
+}
+
+// ValidateUpkeepKey returns true if the types.UpkeepKey is valid, false otherwise
+func (e *encoder) ValidateUpkeepKey(key types.UpkeepKey) (bool, error) {
+	blockKey, upkeepIdentifier, err := e.SplitUpkeepKey(key)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = e.ValidateBlockKey(blockKey)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = e.ValidateUpkeepIdentifier(upkeepIdentifier)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// ValidateUpkeepIdentifier returns true if the types.UpkeepIdentifier is valid, false otherwise
+func (e *encoder) ValidateUpkeepIdentifier(identifier types.UpkeepIdentifier) (bool, error) {
+	identifierInt, ok := identifier.BigInt()
+	if !ok {
+		return false, fmt.Errorf("upkeep identifier is not a big int")
+	}
+	if identifierInt.Cmp(big.NewInt(0)) == -1 {
+		return false, fmt.Errorf("upkeep identifier is not a positive integer")
+	}
+	return true, nil
+}
+
+// ValidateBlockKey returns true if the types.BlockKey is valid, false otherwise
+func (e *encoder) ValidateBlockKey(key types.BlockKey) (bool, error) {
+	keyInt, ok := key.BigInt()
+	if !ok {
+		return false, fmt.Errorf("block key is not a big int")
+	}
+	if keyInt.Cmp(big.NewInt(0)) == -1 {
+		return false, fmt.Errorf("block key is not a positive integer")
+	}
+	return true, nil
+}
+
+// MakeUpkeepKey creates a new types.UpkeepKey from a types.BlockKey and a types.UpkeepIdentifier
+func (e *encoder) MakeUpkeepKey(blockKey types.BlockKey, upkeepIdentifier types.UpkeepIdentifier) types.UpkeepKey {
+	return chain.UpkeepKey(fmt.Sprintf("%s%s%s", blockKey, separator, string(upkeepIdentifier)))
+}
+
+// SplitUpkeepKey splits a types.UpkeepKey into its constituent types.BlockKey and types.UpkeepIdentifier parts
+func (e *encoder) SplitUpkeepKey(upkeepKey types.UpkeepKey) (types.BlockKey, types.UpkeepIdentifier, error) {
+	if upkeepKey == nil {
+		return nil, nil, fmt.Errorf("%w: missing data in upkeep key", errUpkeepKeyNotParsable)
+	}
+	components := strings.Split(upkeepKey.String(), separator)
+	if len(components) != 2 {
+		return nil, nil, fmt.Errorf("%w: missing data in upkeep key", errUpkeepKeyNotParsable)
+	}
+
+	return chain.BlockKey(components[0]), types.UpkeepIdentifier(components[1]), nil
+}
+
+// EncodeReport encodes a report from a list of upkeep results
+func (e *encoder) EncodeReport(toReport []types.UpkeepResult, _ ...encoders.Config) ([]byte, error) {
+	if len(toReport) == 0 {
+		return nil, nil
+	}
+
+	var baseValuesIdx int
+	for i, rpt := range toReport {
+		if rpt.CheckBlockNumber > uint32(baseValuesIdx) {
+			baseValuesIdx = i
```

Comment:
I find this logic surprising; is `CheckBlockNumber` the block number on which the check was performed? If so, won't the largest `CheckBlockNumber` we encounter always be far greater than the number of `UpkeepResult`s that we pass? I think the main reason I found this surprising was because we're comparing block numbers with slice indices?

Reply:
I'll have to look at this more. I wasn't aware that there was a relationship between the `number` of results and the `CheckBlockNumber` of each.

Reply:
can we please fix this logic. It does not make sense to compare checkBlockNumbers with index values